### PR TITLE
LibJS: Add spec comments to FunctionDeclarationInstantiation

### PR DESCRIPTION
**Pull request by shannonbooth**

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp [hunk 1]
```diff
@@ -371,17 +381,55 @@ ThrowCompletionOr<void> ECMAScriptFunctionObject::function_declaration_instantia
...
     }
 
-    auto arguments_object_needed = m_might_need_arguments_object;
+    // 7. Let simpleParameterList be IsSimpleParameterList of formals.
+    bool const simple_parameter_list = has_simple_parameter_list();
 
-    if (this_mode() == ThisMode::Lexical)
-        arguments_object_needed = false;
+    // NOTE: Already set above.
+    // 8. Let hasParameterExpressions be ContainsExpression of formals.
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp:388]: Please swap, spec first NOTE second

File: Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp [hunk 2]
```diff
@@ -371,17 +381,55 @@ ThrowCompletionOr<void> ECMAScriptFunctionObject::function_declaration_instantia
...
 
-    if (parameter_names.contains(vm.names.arguments.as_string()))
-        arguments_object_needed = false;
+    // NOTE: Not needed as we use iteration helpers for this instead.
+    //
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp:391]: Likewise, + no blank line

File: Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp [hunk 3]
```diff
@@ -371,17 +381,55 @@ ThrowCompletionOr<void> ECMAScriptFunctionObject::function_declaration_instantia
...
+    // 11. Let lexicalNames be the LexicallyDeclaredNames of code.
 
+    // 12. Let functionNames be a new empty List.
     HashTable<DeprecatedFlyString> function_names;
+
+    // 13. Let functionsToInitialize be a new empty List.
     Vector<FunctionDeclaration const&> functions_to_initialize;
 
+    // NOTE: This block is done in step 18 below.
+    //
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp:403]: Likewise

File: Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp [hunk 4]
```diff
@@ -371,17 +381,55 @@ ThrowCompletionOr<void> ECMAScriptFunctionObject::function_declaration_instantia
...
+        arguments_object_needed = false;
+    }
+    // 17. Else if parameterNames contains "arguments", then
+    else if (parameter_names.contains(vm.names.arguments.as_string())) {
+        // a. Set argumentsObjectNeeded to false.
+        arguments_object_needed = false;
+    }
+
+    // NOTE: The block below is a combination of step 14 and step 18.
+    //
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp:429]: Likewise

File: Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp [hunk 5]
```diff
@@ -407,42 +455,98 @@ ThrowCompletionOr<void> ECMAScriptFunctionObject::function_declaration_instantia
...
+    // 23. Else,
+    else {
+        // a. Let parameterBindings be parameterNames.
+    }
 
-    // We now treat parameterBindings as parameterNames.
+    // NOTE: We now treat parameterBindings as parameterNames.
 
-    // The spec makes an iterator here to do IteratorBindingInitialization but we just do it manually
+    // NOTE: The spec makes an iterator here to do IteratorBindingInitialization but we just do it manually
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp:544]: Likewise

File: Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp [hunk 6]
```diff
@@ -507,12 +611,25 @@ ThrowCompletionOr<void> ECMAScriptFunctionObject::function_declaration_instantia
     if (scope_body)
         instantiated_var_names.ensure_capacity(scope_body->var_declaration_count());
 
+    // 27. If hasParameterExpressions is false, then
     if (!has_parameter_expressions) {
+        // a. NOTE: Only a single Environment Record is needed for the parameters and top-level vars.
+
+        // NOTE: Done in implementation of step 27.c.i.1 below
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp:618]: Likewise

**Review by linusg — COMMENTED** [review]
Nice one, couple of nitpicks

File: Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp (hunk 1 above)
Comment (shannonbooth) [reply Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp:388]: Thanks! And yeah, I was actually thinking about what style to use for some of them, especially the multi-line ones. Will fix, good to get clarification for future :^)

**Pull request merged**